### PR TITLE
Support train onnx model with tensorflow

### PR DESCRIPTION
**Pull request by chudegao**
1. Update backend.prepare() to generate tf_rep for model training: generate tensor_dict and graph only in training mode. 
2. Add an example to train a onnx model : example/train_onnx_model.py
3. Batchnomalization OP for training model is not supported so far.

### CODE REVIEW

**Comment by chinhuang007** [comment]
@chudegao Thanks for adding this feature! Please sign CLA and clean up code as reported by lgtm, https://lgtm.com/projects/g/onnx/onnx-tensorflow/rev/pr-22b925b7906cedff0b3228f32540fc8ff406a49b

File: onnx_tf/backend.py [hunk 1]
```diff
@@ -107,6 +110,8 @@ def _onnx_graph_to_tensorflow_rep(cls, graph_def, opset, strict, **kwargs):
     # User provided input tensors, in the case the model inputs have unknown shapes
     input_tensor_dict = kwargs[
         'input_tensor_dict'] if 'input_tensor_dict' in kwargs else dict()
+    training_mode = kwargs[
+        'training_mode'] if 'training_mode' in kwargs else dict()
```
Comment (chinhuang007) [inline onnx_tf/backend.py:114]: Looks like training_mode is a bool flag so the default should be false.

File: example/train_onnx_model.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                   tf_rep.tensor_dict['y_truth']), tf.float32))
+
+      train_data, test_data = get_dataset()
+      sess.run(tf_compat.global_variables_initializer())
+      print("==> Train the model..")
+
+      for epoch in range(1, epochs + 1):
+        step = 1
+        next_batch = tf.compat.v1.data.make_one_shot_iterator(
+            train_data).get_next()
```
Comment (chinhuang007) [inline example/train_onnx_model.py:126]: Got errors below, any idea how to fix it? Does it require minimum free memory?
2021-02-04 00:35:22.401127: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 1228800000 exceeds 10% of free system memory.
AttributeError: 'BatchDataset' object has no attribute '_make_one_shot_iterator'

File: onnx_tf/backend.py (hunk 1 above)
Comment (chudegao) [reply onnx_tf/backend.py:114]: Update to False as default value.

File: example/train_onnx_model.py (hunk 2 above)
Comment (chudegao) [reply example/train_onnx_model.py:126]: @chinhuang007 May I know your tensorflow version?

Comment (chudegao) [reply example/train_onnx_model.py]: @chinhuang007 I did not hit this issue with tf2.4 and tf2.3.1 in both mac and linux environment. What's your test environment?

Comment (chinhuang007) [reply example/train_onnx_model.py]: Python version:
3.6.9 (default, Nov  7 2019, 10:44:02)
[GCC 8.3.0]
ONNX version:
1.8.0
ONNX-TF version:
1.7.0
Tensorflow version:
2.4.0

File: onnx_tf/backend.py [hunk 3]
```diff
@@ -254,12 +279,13 @@ def validate_initializer_name(name):
...
           ":", "_tf_") + "_" + get_unique_suffix() if ":" in name else name
 
-    return [(init.name,
-             tf.constant(tensor2list(init),
-                         shape=init.dims,
-                         dtype=data_type.onnx2tf(init.data_type),
-                         name=validate_initializer_name(init.name)))
-            for init in initializer]
+    tensor_dict = [(init.name,
+                    tf.Variable(np.array(tensor2list(init)).reshape(init.dims),
```
Comment (chinhuang007) [inline onnx_tf/backend.py:283]: The initializers should be constant if training_mode is False. For inference, they are not going to change.

Comment (chudegao) [reply onnx_tf/backend.py]: Yes.  Fixed it.

File: example/train_onnx_model.py (hunk 2 above)
Comment (chudegao) [reply example/train_onnx_model.py]: I updated cifar10 dataset to mnist, the memory requirement should be reduced from 15G to 5G.

**Review by chinhuang007 — APPROVED** [review]
LGTM!

**Pull request merged**